edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs

```diff
@@ -0,0 +1,812 @@
+                                this.operatorWatch.ForEach(watch => watch.Dispose());
+                                this.operatorWatch = Option.None<Watcher<object>>();
+
+                                // kick off a new watch
+                                this.client.ListClusterCustomObjectWithHttpMessagesAsync(
+                                    Constants.K8sCrdGroup,
+                                    Constants.K8sApiVersion,
+                                    Constants.K8sCrdPlural,
+                                    watch: true).ContinueWith(this.ListCrdComplete);
+                            },
+                            onError: this.logger.ExceptionInCustomResourceWatch
+                        ));
+                }
+                else
+                {
+                    this.logger.NullListResponse("ListClusterCustomObjectWithHttpMessagesAsync", "http response");
+                    throw new NullReferenceException("Null response from ListClusterCustomObjectWithHttpMessagesAsync");
+                }
+            }
+            else
+            {
+                this.logger.NullListResponse("ListClusterCustomObjectWithHttpMessagesAsync", "task");
+                throw new NullReferenceException("Null Task from ListClusterCustomObjectWithHttpMessagesAsync");
+            }
+        }
+
+        private async Task WatchDeploymentEventsAsync(WatchEventType type, object custom)
+        {
+            using (await this.watchLock.LockAsync())
+            {
+                EdgeDeploymentDefinition<TConfig> customObject;
+                try
+                {
+                    string customString = JsonConvert.SerializeObject(custom);
+                    customObject = this.deploymentSerde.Deserialize(customString);
+                }
+                catch (Exception e)
+                {
+                    this.logger.EdgeDeploymentDeserializeFail(e);
+                    return;
+                }
+
+                // only operate on the device that matches this operator.
+                if (String.Equals(customObject.Metadata.Name, this.resourceName, StringComparison.OrdinalIgnoreCase))
+                {
+                    V1ServiceList currentServices = await this.client.ListNamespacedServiceAsync(KubeUtils.K8sNamespace, labelSelector: this.deploymentSelector);
+                    V1DeploymentList currentDeployments = await this.client.ListNamespacedDeploymentAsync(KubeUtils.K8sNamespace, labelSelector: this.deploymentSelector);
+                    this.logger.DeploymentStatus(type, this.resourceName);
+                    switch (type)
+                    {
+                        case WatchEventType.Added:
+                        case WatchEventType.Modified:
+                            this.ManageDeployments(currentServices, currentDeployments, customObject);
+                            break;
+
+                        case WatchEventType.Deleted:
```

Comment:
>WatchEventType.Deleted [](start = 29, length = 22)

Since you are maintaining a list of current modules, should you clean that up as well?

Reply:
I am now clearing the list of modules.